Support importing compressed EC keys in webcrypto

angelaizg opened the pull request:
#### c2800dd46bb2deeea08659eb0d3ae6d9eab219ac
<pre>
Support importing compressed EC keys in webcrypto
<a href="https://bugs.webkit.org/show_bug.cgi?id=244032">https://bugs.webkit.org/show_bug.cgi?id=244032</a>
rdar://86211772

Reviewed by Youenn Fablet

The compressed point format includes the least significant bit of the second coordinates in the first byte to know the orientation , so that is why the index is incremented one unit when dealing with compressed curves .  Otherwise the function ccec_compact_import_pub_size receivesthe wrong number, expecting always a number greater in just one unit.
* LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384-expected.txt: Added.
* LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384.html: Added.
* LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PAL/pal/spi/cocoa/CommonCryptoSPI.h:
* Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
(WebCore::CryptoKeyEC::platformImportSpki):

Canonical link: <a href="https://commits.webkit.org/254310@main">https://commits.webkit.org/254310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40f7b1078e277809614fdc967024bb83f168711d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88582 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33144 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97774 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/153669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92578 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31636 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27212 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80807 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92410 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94212 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25095 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75514 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25038 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79980 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80053 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68001 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29280 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14049 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29166 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15064 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3040 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37994 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31291 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34173 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,16 +251,27 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-
-    return create(identifier, curve, CryptoKeyType::Public, PlatformECKeyContainer(ccPublicKey), extractable, usages);
+        
+    if ( uncompressed ){
```
Coding style probably tells us to write it as `if (uncompressed) {`

angelaizg replied:
I changed  LayoutTests/platform/glib/TestExpectations as the fix was only made in CryptoKeyECMac

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,16 +251,27 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-    return create(identifier, curve, CryptoKeyType::Public, PlatformECKeyContainer(ccPublicKey), extractable, usages);
+        
+    if ( uncompressed ){
+    if ( CCECCryptorImportKey(kCCImportKeyBinary, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
```
Needs some space to increment properly.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,16 +251,27 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+    if( compressed ){
+    if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
+          return nullptr;
+      }
```
With the way it is currently written, if initialOctet is 1, we would probably use a nullptr public key.
In the past code path, we would try to use kCCImportKeyBinary.
Either we need to return early if octet is not 2 3 or 4.
Or we could write it in the following way:
`if (compressed) {
      use kCCImportKeyCompact
} else {
     use existing code path (kCCImportKeyBinary)`
}

angelaizg replied:
Already changed to check if it is compressed or not and then use CCECCryptorImportKey with either  kCCImportKeyBinary or kCCImportKeyCompact.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,16 +251,27 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+    index += 1;
+        
+        
+    bool uncompressed = initialOctet == 4 or initialOctet == 04 ;
```
Why testing 4 and 04?
Space at the end of the line is unneeded.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,16 +251,27 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+        
+        
+    bool uncompressed = initialOctet == 4 or initialOctet == 04 ;
+    bool compressed = initialOctet == 03 or initialOctet == 02;
```
Can we change 03 to 3? Ditto for 02?

youennf reviewed (changes requested):
As Alexey said, we probably need some changes to -expected.txt in this patch, hopefully we have some WPT tests in that area.
Also, it is interesting to look at test results in https://ews-build.webkit.org/#/builders/70/builds/13504.
Maybe the crashes are due to using a nullptr key?

panva commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,16 +251,28 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+    index += 1;
+        
+        
+    bool uncompressed = initialOctet == 4 ;
```
```suggestion
    bool uncompressed = initialOctet == 4;
```

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,15 +252,18 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-    if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index))
-        return nullptr;
-
+    index += bytesUsedToEncodedLength(keyData[index])+1; // Read length
```
s/+1/ + 1/

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,15 +252,18 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-        return nullptr;
-
+    if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index)) {
+        if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index+1, keyData.size() - index-1, ccECKeyPublic, &ccPublicKey)) {
```
Can you investigate why we need the + 1 here.
Ideally the commit message would indicate why we need it and refer to a spec.
I would tend to write it as:
```
++index;
if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index, keyData.size() .- index, ...))
```

angelaizg replied:
The compressed point format includes the least significant bit of the second coordinates in the first byte to know the orientation , so I think that is why we need one more. I have been doing some debugging and found out that otherwise the function `ccec_compact_import_pub_size` the wrong  number, expecting always a number greater in just one unit.

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,15 +252,18 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+        if (CCECCryptorImportKey(kCCImportKeyCompact, keyData.data() + index+1, keyData.size() - index-1, ccECKeyPublic, &ccPublicKey)) {
+            return nullptr;
+            
+        }
```
The '{' and '}' are not needed for if blocks if the block is a one liner.

youennf commented on LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384.html:
```diff
@@ -0,0 +1,38 @@
+<div id="console"></div>
+
+<script>
+description("Test importing a P-384 SPKI ECDSA COMPRESSED key ");
```
I am surprised there are not related tests in LayoutTests/imported/w3c/web-platform-tests.
Does it mean Chrome and Firefox do not support uncompressed keys?

youennf commented on Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
```diff
@@ -252,15 +252,18 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
-    if (CCECCryptorImportKey(kCCImportKeyBinary, keyData.data() + index, keyData.size() - index, ccECKeyPublic, &ccPublicKey))
-        return nullptr;
-
+    if (!doesUncompressedPointMatchNamedCurve(curve, keyData.size() - index)) {
```
I would tend to write if as `if (doesUncompressedPointMatchNamedCurve())` instead of if (! doesUncompressedPointMatchNamedCurve())`

zdobersek reviewed (approved):
There's the empty `ecdsa-import-compressed-spki-key-p384.txt` in the commit, don't think that's necessary?

Commit:
Support importing compressed EC keys in webcrypto
https://bugs.webkit.org/show_bug.cgi?id=244032
rdar://86211772

Reviewed by Youenn Fablet

The compressed point format includes the least significant bit of the second coordinates in the first byte to know the orientation , so that is why the index is incremented one unit when dealing with compressed curves .  Otherwise the function ccec_compact_import_pub_size receivesthe wrong number, expecting always a number greater in just one unit.
* LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384-expected.txt: Added.
* LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384.html: Added.
* LayoutTests/crypto/subtle/ecdsa-import-compressed-spki-key-p384.txt: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/PAL/pal/spi/cocoa/CommonCryptoSPI.h:
* Source/WebCore/crypto/mac/CryptoKeyECMac.cpp:
(WebCore::CryptoKeyEC::platformImportSpki):

Canonical link: https://commits.webkit.org/254310@main
